File: flutter-tela-pokemon/lib/widgets/type_chip.dart
import 'package:flutter/material.dart';

import '../theme/pokemon_types.dart';

/// Tamanhos previstos para o chip de tipo.
enum TypeChipSize { small, large }

/// Chip colorido de tipo de Pokémon.
///
/// Equivalente ao componente `TypeChip` da versão em React Native: é o menor
/// bloco reutilizável da tela e aparece tanto no cabeçalho (variante invertida)
/// quanto em cada linha da lista de movimentos.
class TypeChip extends StatelessWidget {
  const TypeChip({
    super.key,
    required this.typeName,
    this.size = TypeChipSize.small,
    this.inverted = false,
  });

  final String typeName;
  final TypeChipSize size;

  /// Quando verdadeiro, o chip fica branco com o texto na cor do tipo — usado
  /// sobre o cabeçalho colorido.
  final bool inverted;

  @override
  Widget build(BuildContext context) {
    final PokemonTypeInfo info = typeInfo(typeName);
    final bool isLarge = size == TypeChipSize.large;

    final Color background = inverted ? Colors.white : info.color;
    final Color foreground =
        inverted ? info.color : contrastingTextColor(info.color);

    return Container(
      padding: EdgeInsets.symmetric(
        horizontal: isLarge ? 16 : 10,
        vertical: isLarge ? 8 : 4,
      ),
      decoration: BoxDecoration(
        color: background,
        borderRadius: BorderRadius.circular(999),
      ),
      child: Text(
        info.label,
        style: TextStyle(
          color: foreground,
          fontSize: isLarge ? 14 : 12,
          fontWeight: FontWeight.w700,
        ),
      ),
    );
  }
}

File: react-native-pokemon-explorer/src/components/TypeChip.js
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

import { radius } from '../theme/colors';
import { getContrastingTextColor, getTypeInfo } from '../theme/pokemonTypes';

/**
 * Chip colorido de tipo de Pokémon.
 *
 * Componente base reaproveitado nos cards da listagem, no cabeçalho do detalhe
 * e em cada linha da lista de movimentos.
 *
 * variant "solid"    -> fundo na cor do tipo (cards e lista de movimentos)
 * variant "inverted" -> fundo branco com texto na cor do tipo (cabeçalho colorido)
 */
export default function TypeChip({ type, size = 'md', variant = 'solid' }) {
  const info = getTypeInfo(type);
  const isSolid = variant === 'solid';

  const backgroundColor = isSolid ? info.color : '#FFFFFF';
  const textColor = isSolid ? getContrastingTextColor(info.color) : info.color;

  return (
    <View style={[styles.chip, styles[`${size}Chip`], { backgroundColor }]}>
      <Text style={[styles.label, styles[`${size}Label`], { color: textColor }]} numberOfLines={1}>
        {info.label}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  chip: {
    borderRadius: radius.pill,
    alignSelf: 'flex-start',
  },
  label: {
    fontWeight: '700',
  },
  smChip: { paddingHorizontal: 9, paddingVertical: 3 },
  smLabel: { fontSize: 11 },
  mdChip: { paddingHorizontal: 12, paddingVertical: 5 },
  mdLabel: { fontSize: 12.5 },
  lgChip: { paddingHorizontal: 16, paddingVertical: 8 },
  lgLabel: { fontSize: 14 },
});
